File: light-control/algorithms/lightning/lightning.py
import logging
import math
import time
import random
from .config import config
from .._meta import BaseAlgorithm
from .._meta.pixel import Pixel
from .._meta.pulse import Pulse

class Algorithm(BaseAlgorithm):
    def __init__(self, num_pixels, settings) -> None:
        super().__init__(num_pixels, config, settings)
        
        self.bolts = []
        self.flash_color = Pixel(*settings['color'])
        self.max_bolts = settings['max_bolts']
        self.bolt_prob = settings['bolt_prob']
        self.bump_prob = settings['bump_prob']

        self.bump_perc = 0.3
    
    def add_bolt(self):
        if len(self.bolts) < self.max_bolts and random.random() < self.bolt_prob:
            position = random.randrange(0, len(self.pixels))
            width = random.randrange(4, 20)
            self.bolts.append({
                "pulse": Pulse(position, Pixel(self.flash_color.hue, self.flash_color.sat, self.flash_color.val), width, 4, 4),
                "percentage": 1.0
            })
    
    def fade_bolts(self, elapsed_seconds):
        for bolt in self.bolts:
            if (random.random() < self.bump_prob):
                bolt['percentage'] = min(1.0, bolt['percentage'] + self.bump_perc)
            else:
                bolt['percentage'] = max(0, bolt['percentage'] - elapsed_seconds * 0.5)
            bolt['pulse'].color.val = bolt['percentage'] * self.flash_color.val
    
    def clear_bolts(self):
        self.bolts = list(filter(lambda bolt: bolt['percentage'] > 0, self.bolts))

    def run_cycle(self, elapsed_millis, elapsed_seconds):
        for pixel in self.pixels:
            pixel.clear()
        self.add_bolt()

        for bolt in self.bolts:
            bolt['pulse'].apply_pulse(self.pixels)

        self.fade_bolts(elapsed_seconds)
        self.clear_bolts()

        return super().run_cycle(elapsed_millis, elapsed_seconds)

File: web/server/src/algorithms/lightning/lightning.ts
import { BaseAlgorithm, Pixel, Pulse } from '../_meta/index'
import { config } from './config'

interface Bolt {
  pulse: Pulse
  percentage: number
}

const BUMP_PERC = 0.3

export class Algorithm extends BaseAlgorithm {
  private bolts: Bolt[] = []
  private readonly flashColor: Pixel
  private readonly maxBolts: number
  private readonly boltProb: number
  private readonly bumpProb: number

  constructor(numPixels: number, settings: Record<string, unknown>) {
    super(numPixels, config, settings)
    const color = settings['color'] as number[]
    this.flashColor = new Pixel(color[0], color[1], color[2], color[3] ?? 0)
    this.maxBolts = settings['max_bolts'] as number
    this.boltProb = settings['bolt_prob'] as number
    this.bumpProb = settings['bump_prob'] as number
  }

  private addBolt(): void {
    if (this.bolts.length < this.maxBolts && Math.random() < this.boltProb) {
      const position = Math.floor(Math.random() * this.pixels.length)
      const width = 4 + Math.floor(Math.random() * 16)
      this.bolts.push({
        pulse: new Pulse(
          position,
          new Pixel(this.flashColor.hue, this.flashColor.sat, this.flashColor.val),
          width,
          4,
          4,
        ),
        percentage: 1.0,
      })
    }
  }

  private fadeBolts(elapsedSeconds: number): void {
    for (const bolt of this.bolts) {
      if (Math.random() < this.bumpProb) {
        bolt.percentage = Math.min(1.0, bolt.percentage + BUMP_PERC)
      } else {
        bolt.percentage = Math.max(0, bolt.percentage - elapsedSeconds * 0.5)
      }
      bolt.pulse.color.val = bolt.percentage * this.flashColor.val
    }
  }

  private clearBolts(): void {
    this.bolts = this.bolts.filter(bolt => bolt.percentage > 0)
  }

  runCycle(elapsedMillis: number, elapsedSeconds: number): boolean {
    for (const pixel of this.pixels) pixel.clear()

    this.fadeBolts(elapsedSeconds)
    this.clearBolts()
    this.addBolt()

    for (const bolt of this.bolts) {
      bolt.pulse.applyPulse(this.pixels)
    }

    return super.runCycle(elapsedMillis, elapsedSeconds)
  }
}
